007-design-docs.md

```diff
@@ -0,0 +1,97 @@
+
+### Where to write a design document
+
+A design document should be a markdown file located in the `design` directory of the `strimzi-kafka-operator` project.
+Design documents for other Strimzi projects such as Oauth should also be located here as the intent is that the design documents are all colocated and that all components are a part of 'the operator' project.
+
+### When is a design document complete
+
+A design document is considered complete when the reviewers and maintainers believe the document to accurately reflect the current implementation of the components behaviour. Once approved the document is merged and is understood to be correct, any mistakes or changes to behaviour should be corrected in the form of a pull request.
+
+### How to write a design document
+
+The design documents would live in the strimzi-kafka-operator repo and would match the current code state, so that code changes could be audited with their corresponding design doc changes.
+
+A design document would be split into `X` parts
+
+**Title**
+
+The title of the design doc, normally this would pertain to a single component or one larger aspect of the whole of the Strimzi project. New pages would be discussed and decided on as part of a proposal where needed.
+
+**Overview**
+
+A high level summary that any user of Strimzi should be able understand and use to decide if it’s useful for them to read the rest of the doc. It should be 3 paragraphs max.
+
+**Context**
+
+A description of the problem the program/code tackles, why the program/code is necessary, what people need to know to assess this project, and how it fits into the technical strategy of Strimzi.
+
+**Goals and Non-Goals**
+
+The Goals section should describe the user-driven impact of the code and specify how to measure success using metrics, normally that would be tested.
+Non-Goals are equally important to describe which problems you won’t be fixing so everyone is on the same page.
+
+**Milestones**
+
+A list of measurable checkpoints, if there is future progress planned, including a milestone for the future in the design doc maye be useful for a reader to understand where the project is heading.
+
+**High Level Description**
+
+High level description describing the current implementation as well as example flow(s) to illustrate how users interact with this system and/or how data flow through it. This should build on the context.
+
+**Technical Architecture section**
+
+Walk through the flows in great detail to concertise the implementation. Feel free to include diagrams where relevant and sub-sections for edge-cases and implementation details.
+
+## Summary of changes
+
+Alongside code changes, behavioural changes of the operator will also be documented in the above described design documentation format.
+
+Newly contributed proposals or issues would describe how their changes might alter the design documentation and are expected to be delivered alongside their code changes.
```

Comment:
This is hard given no previous design docs for the old features.

Reply:
True, but you can deploy that argument against any attempt to improve the status quo.

I think all we really want is a way to improve from where we are (and asymptotically reach design documentation nirvana). That doesn't require placing the entire burden on the first person to come along who wants to make a change in a given area. I would avoid including this language here, since it's another barrier, and instead trust that as part of the process we remember to ask people nicely to improve on what we have.